### PR TITLE
Fix some unexpected_cfg lint warnings

### PR DESCRIPTION
**Pull request by Manishearth**
Latest Rust has a bunch of warnings around unexpected `cfg` keys, to avoid mistakes and encourage use of features. This can be worked around by documenting deliberate CFG keys in Cargo.toml (or by emitting them from the build script, which I'd rather not introduce).

This doesn't fix all of them, we have a *ton* of `cfg(skip)` that we should fix. Perhaps `cfg(not(any()))` for now, though really we should tweak the ctor macros to not need this, the cfg is a hack.

### CODE REVIEW

**Comment by Manishearth** [comment]
> This doesn't fix all of them, we have a _ton_ of `cfg(skip)` that we should fix. Perhaps `cfg(not(any()))` for now, though really we should tweak the ctor macros to not need this, the cfg is a hack.

Gonna make this change in this PR to try and get CI passing again. We can figure out a longer term macro solution later: this is extremely hacky and works in an unintuitive way.

File: components/calendar/src/any_calendar.rs [hunk 1]
```diff
@@ -758,7 +758,7 @@ impl AnyCalendar {
         locale: include,
         options: skip,
         error: DataError,
-        #[cfg(skip)]
+        #[cfg(not(all()))]
```
Comment (sffc) [inline components/calendar/src/any_calendar.rs:761]: Suggestion: these `cfg(skip)`s in the invocations of `gen_any_buffer_data_constructors!` could be better modeled by changing the macro itself to not emit these functions.

File: tools/bakeddata-scripts/template/Cargo.toml.template [hunk 2]
```diff
@@ -21,4 +21,8 @@ cldr = { tagged = "_cldr_tag_" }
 icuexport = { tagged = "_icuexport_tag_" }
 segmenter_lstm = { tagged = "_segmenter_lstm_tag_" }
 
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(icu4x_custom_data)'] }
+
 [dependencies]
+
```
Comment (sffc) [inline tools/bakeddata-scripts/template/Cargo.toml.template:28]: Nit: Extra newline

File: tools/bakeddata-scripts/template/Cargo.toml.template [hunk 3]
```diff
@@ -21,4 +21,8 @@ cldr = { tagged = "_cldr_tag_" }
 icuexport = { tagged = "_icuexport_tag_" }
 segmenter_lstm = { tagged = "_segmenter_lstm_tag_" }
 
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(icu4x_custom_data)'] }
```
Comment (sffc) [inline tools/bakeddata-scripts/template/Cargo.toml.template:25]: Question: Why level = warn?

File: components/calendar/src/any_calendar.rs (hunk 1 above)
Comment (Manishearth) [reply components/calendar/src/any_calendar.rs:761]: Yes, that's what I said in the comments, but I didn't want to figure that out whilst fixing nightly CI.

File: tools/bakeddata-scripts/template/Cargo.toml.template (hunk 3 above)
Comment (Manishearth) [reply tools/bakeddata-scripts/template/Cargo.toml.template:25]: Oh I copied this from the error message. We don't need it.

**Comment by Manishearth** [comment]
> [cargo-make] WARN - Unable to parse Cargo.toml via cargo-metadata, fallbacking. `cargo metadata` exited with an error: error: missing field `level`

It appears we are required to include the lint level. Hm.

**Comment by sffc** [comment]
Can the level be "notice" or "ignore" or something?

**Comment by sffc** [comment]
We definitely want the lint for unknown directives. The suppression is only for the few we include intentionally.

**Comment by Manishearth** [comment]
@sffc I don't understand, I had previously set the lint to "warn" and you didn't want that, what is the difference between "warn" and what you mean by "notice"?

**Comment by Manishearth** [comment]
This fixes all but one case, and that needs to be fixed by improving the macro first. I'm working on that as a separate PR for now.

File: provider/data/experimental/Cargo.toml [hunk 4]
```diff
@@ -21,5 +21,9 @@ cldr = { tagged = "45.0.0" }
 icuexport = { tagged = "icu4x/2024-05-16/75.x" }
 segmenter_lstm = { tagged = "v0.1.0" }
 
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(icu4x_custom_data)'] }
+
 [dependencies]
+
```
Comment (robertbastian) [inline provider/data/experimental/Cargo.toml:28]: newline

File: tools/bakeddata-scripts/template/Cargo.toml.template (hunk 3 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply tools/bakeddata-scripts/template/Cargo.toml.template:25]: It needs to be unconditionally emitted with a "this cfg exists" directive

[the remark this answers is not in the document]
Comment (robertbastian) [reply tools/bakeddata-scripts/template/Cargo.toml.template:25]: Can this be removed now?

**Comment by Manishearth** [comment]
I resolved the comments but I also undid the cfg(skip) changes when fixing merge conflicts since that's going to conflict more.

**Review by robertbastian — APPROVED** [review]

**Pull request merged**